### ISSUE TITLE
Helm does not wait for Statefulset to be ready after upgrade

### ISSUE THREAD

**Issue by libsamek** [helm/helm#10163]
Helm does not wait for Statefulset to be ready after upgrade to 3.6.3 from 3.3.1. I have tried using versions:
* 3.7.0 (does not work),
* 3.6.3 (does not work),
* 3.5.4 (works!),
* 3.3.1 (works!).

It seems that this behavior was introduced with: https://github.com/helm/helm/commit/98d98369ed9abcaec9b75352e07dd6f759a8804f

Helm 3.7.0 behavior - unexpected:

```
helm upgrade --install --reset-values XXX chart --version X.Y.Z --wait --timeout 900s --create-namespace --namespace dummy --values /tmp/helmfile351904154/XXX-values-6fd57cf55 ... --history-max 10 --debug
helm:whTHc> history.go:56: [debug] getting history for release XXX
helm:whTHc> upgrade.go:139: [debug] preparing upgrade for XXX
helm:whTHc> upgrade.go:488: [debug] resetting values to the chart's original version
helm:whTHc> upgrade.go:147: [debug] performing update for XXX
helm:whTHc> upgrade.go:319: [debug] creating upgraded release for XXX
helm:whTHc> client.go:218: [debug] checking 4 resources for changes
helm:whTHc> client.go:501: [debug] Looks like there are no changes for ConfigMap "XXX"
helm:whTHc> client.go:501: [debug] Looks like there are no changes for Service "XXX"
helm:whTHc> wait.go:48: [debug] beginning wait for 4 resources with timeout of 15m0s
helm:whTHc> upgrade.go:154: [debug] updating status for upgraded release for XXX
helm:whTHc> Release "XXX" has been upgraded. Happy Helming!
NAME: XXX
...
```

Helm 3.5.4 behavior - expected:
```
exec: helm upgrade --install --reset-values XXX chart --version X.Y.Z --wait --timeout 900s --create-namespace --namespace dummy --values /tmp/helmfile549477625/XXX-values-68c7579fb5 ... --history-max 10 --debug
helm:whTHc> history.go:56: [debug] getting history for release XXX
helm:whTHc> upgrade.go:123: [debug] preparing upgrade for XXX
helm:whTHc> upgrade.go:437: [debug] resetting values to the chart's original version
helm:whTHc> upgrade.go:131: [debug] performing update for XXX
helm:whTHc> upgrade.go:303: [debug] creating upgraded release for XXX
helm:whTHc> client.go:201: [debug] checking 4 resources for changes
helm:whTHc> client.go:464: [debug] Looks like there are no changes for PodDisruptionBudget "XXX"
helm:whTHc> client.go:464: [debug] Looks like there are no changes for ConfigMap "XXX"
helm:whTHc> client.go:464: [debug] Looks like there are no changes for Service "XXX"
helm:whTHc> wait.go:53: [debug] beginning wait for 4 resources with timeout of 15m0s
helm:whTHc> wait.go:348: [debug] StatefulSet is not ready: dummy/XXX. 0 out of 2 expected pods have been scheduled
helm:whTHc> wait.go:348: [debug] StatefulSet is not ready: dummy/XXX. 0 out of 2 expected pods have been scheduled
helm:whTHc> wait.go:348: [debug] StatefulSet is not ready: dummy/XXX. 0 out of 2 expected pods have been scheduled
helm:whTHc> wait.go:348: [debug] StatefulSet is not ready: dummy/XXX. 1 out of 2 expected pods have been scheduled
helm:whTHc> wait.go:348: [debug] StatefulSet is not ready: dummy/XXX. 1 out of 2 expected pods have been scheduled
...
helm:whTHc> wait.go:353: [debug] StatefulSet is not ready: dummy/XXX. 1 out of 2 expected pods are ready
helm:whTHc> wait.go:353: [debug] StatefulSet is not ready: dummy/XXX. 1 out of 2 expected pods are ready
...
helm:whTHc> upgrade.go:138: [debug] updating status for upgraded release for XXX
helm:whTHc> Release "XXX" has been upgraded. Happy Helming!
NAME: XXX
```

It's weird because I have similar releases of this chart where `--wait` works as expected. I do not know if it is relevant but release XXX was converted from Helm 2 release using 2to3 plugin a while ago.

Output of `helm version`:
version.BuildInfo{Version:"v3.7.0", GitCommit:"eeac83883cb4014fe60267ec6373570374ce770b", GitTreeState:"clean", GoVersion:"go1.16.8"}

Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.10", GitCommit:"8152330a2b6ca3621196e62966ef761b8f5a61bb", GitTreeState:"clean", BuildDate:"2021-08-11T18:06:15Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.9", GitCommit:"7a576bc3935a6b555e33346fd73ad77c925e9e4a", GitTreeState:"clean", BuildDate:"2021-07-15T20:56:38Z", GoVersion:"go1.15.14", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
AWS+Kops

**Comment by bacongobbler** [comment]
Hi @libsamek. Thanks for looking into this.

Can you please provide an example to be used for testing? We need a way to reproduce the issue to try and determine the proper cause.

**Comment by libsamek** [comment]
Hey @bacongobbler, it will probably take a couple of days to prepare an example. I'm not sure what is causing this because in most cases this is not an issue.

**Comment by libsamek** [comment]
Good idea! I was thinking of doing something like that myself. I'll try reproducing this with a chart I can share. Currently, it's not a priority because we encountered this issue only in one case with StatefulSet. Anyhow, I'll try to find some time to reproduce this.

**Comment by libsamek** [comment]
@dnwe, it seems that your fix is working! 🎉 

It seems there is a race condition as you have assumed.

So, what do we need to do to get this merged into master?

Output:
```
helm:gbaiC> history.go:56: [debug] getting history for release XXX
helm:gbaiC> upgrade.go:139: [debug] preparing upgrade for XXX
helm:gbaiC> upgrade.go:491: [debug] resetting values to the chart's original version
helm:gbaiC> upgrade.go:147: [debug] performing update for XXX
helm:gbaiC> upgrade.go:319: [debug] creating upgraded release for XXX
helm:gbaiC> client.go:218: [debug] checking 4 resources for changes
helm:gbaiC> client.go:501: [debug] Looks like there are no changes for ConfigMap "XXX"
helm:gbaiC> client.go:501: [debug] Looks like there are no changes for Service "XXX"
helm:gbaiC> upgrade.go:383: [debug] waiting for release XXX resources (created=0, updated=4, deleted=0)
wait.go:48: [debug] beginning wait for 4 resources with timeout of 15m0s
helm:gbaiC> ready.go:362: [debug] StatefulSet is not ready: develop/XXX. update has not yet been observed
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 0 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods have been scheduled
...
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:388: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods have been scheduled
helm:gbaiC> ready.go:393: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods are ready
helm:gbaiC> ready.go:393: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods are ready
helm:gbaiC> ready.go:393: [debug] StatefulSet is not ready: develop/XXX. 1 out of 2 expected pods are ready
...
helm:gbaiC> ready.go:402: [debug] StatefulSet is ready: develop/XXX. 2 out of 2 expected pods are ready
helm:gbaiC> upgrade.go:154: [debug] updating status for upgraded release for XXX
helm:gbaiC> Release "XXX" has been upgraded. Happy Helming!
NAME: XXX
LAST DEPLOYED: Mon Nov 22 15:13:03 2021
```

**Comment by libsamek** [comment]
@bacongobbler I have also prepared a repository with information on how to reproduce this: https://github.com/libsamek/dummy.

Please let me know if this is sufficient.

**Comment by jonny08152** [comment]
Any news on this?

We see this issue as well and it hinders us from upgrading helm to the newer (>v3.5.4) versions. The issue seems to be a race condition (possibly introduced in [98d9836](https://github.com/helm/helm/commit/98d98369ed9abcaec9b75352e07dd6f759a8804f)) as this only happens when running helm in-cluster when the kube api has really short response times. It is easily triggered by introducing and removing an artificial delay to the kube-api.

I can provide detailed information and a testing scenario to reproduce this if necessary.

The [fix](https://github.com/helm/helm/pull/10622) of @dnwe works great. It just misses some reviews from the maintainers to be merged back.

**Comment by libsamek** [comment]
Any news on this?

We are in a similar position as mentioned by @jonny08152, we cannot upgrade Helm without this fix.

I am also prepared to help with testing if needed.

**Comment by dnwe** [comment]
Fixed in [Helm v3.9.1](https://github.com/helm/helm/releases/tag/v3.9.1)
